### ISSUE TITLE
Copy Paste from editor over SSH - OSC 52 support

### ISSUE THREAD

**Issue by luzik** [gnunn1/tilix#2198]
As more and more console editors make use of "OSC 52" please consider supporting it by tilix

https://github.com/neovim/neovim/pull/25872/commits/748bc4d22d35b40f0795015e5d93113b526adf22

**Comment by egmontkob** [comment]
(I'm not a Tilix developer but I'm a VTE developer.)

Tilix uses the VTE widget for terminal emulation.

VTE developers deliberately refuse to add OSC 52 support because of its known security/privacy problems. At least two protocols that would fix these problems have been proposed, but no one had the resources yet to implement them and start convincing other terminals and apps to follow.

In the mean time, shame on every terminal emulator and every terminal-based application that adds OSC 52 support, despite its well-known security flaws. They should all deliberately reject to implement it, and cooperate to design and drive the adoptation of a new, safe solution. My 2 cents.

**Comment by luzik** [comment]
Thank you for clarification.
As I am aware of security concerns, in my environment they do not apply - I own all the machines involved. It would be awesome to have option like '--enable-unsecure-osc52' until something betters comes in. Especially in one way - editor can write to clipboard without read capabilities - this options seems to be safer.

**Comment by egmontkob** [comment]
I understand you. And I'm still a bit skeptical:

>  in my environment they do not apply - I own all the machines involved

Do you know _for 100% sure_ that you'll never ever ssh / telnet / netcat (even for debugging, maybe on http port 80 or so) to a machine outside of your security realm; or if you do then you'll remember to turn off this feature beforehand?

---

I (or we), as developer(s), have to think differently.

If we offer a run-time switch (visible config option, hidden config option, cmdline switch etc.) then users will find it, users will tell each other on forums to enable that option, and many will blindly enable it without properly considering the security implications. Security is defeated.

If it's a compile-time switch then some distros won't enable it (their users still complaining about the lack of copy-paste support), some other will perhaps enable (posing the security risk to all their users). Neither of which is good.

If end users who know it's safe _for them_ need to recompile the package anyway, then you can go ahead and patch in OSC 52 support, it shouldn't be very hard (well, first check if someone has already done and shared that).

"Let's ship a feature for only those users whose setup meets certain security criteria" is in my opinion not an acceptable option for developers of terminal emulators, nor for developers of terminal-based apps. If people want to have this feature, it's the developer's job to ship a solution that can safely work for all users.

From my point of view, as a developer who contributes improvements, security cannot be optional, security must be mandatory.

Maybe we should just go ahead, design and implement our new protocol, and post feature requests against apps to see if it grains traction...

**Comment by antoine2711** [comment]
I understand the security issue, but a simple solution would be an internal clipbord in each terminal windows, with 2 buttons, In/Out, that would transfert between the system and the internal clipboard. It's simple, yet very pratical.

Can this be considered?

Regards, Antoine

**Comment by egmontkob** [comment]
I have two problems here.

The first is that this would be a terrible user experience.

The second is: Tilix builds on top of VTE, and VTE refuses to add the necessary features due to its security implications. In order to implement this in Tilix, VTE would have to expose a set of features towards Tilix (or whatever other VTE-based terminal app); which, in turn, would probably be misused in a couple of VTE-based terminals and would open up the security problems there.

I personally do not intent to spend any amount of work towards a bad solution.

**Comment by antoine2711** [comment]
> The first is that this would be a terrible user experience.

I've seen this a lot, in Graphic Remote app, like Timbuktu (I used it in 1987 to 2013), some implementation of VNC, etc. It give the possibility of protecting your clipboard while also making it easy to exchange the content between the host and the remote machines. Not unpleasant to use.
 
> The second is: Tilix builds on top of VTE, and VTE refuses to add the necessary features due to its security implications. In order to implement this in Tilix, VTE would have to expose a set of features towards Tilix (or whatever other VTE-based terminal app); which, in turn, would probably be misused in a couple of VTE-based terminals and would open up the security problems there.

Of course, this is a show stopper. And for sure, VTE would have to document explicitly the security concerns to the developers.

Thanks for taking the time to answer me, regards, Antoine

**Comment by egmontkob** [comment]
> I've seen this a lot, in Graphic Remote app

Fair point, I vaguely recall (from ages ago) that VMware had such menu entries (of course I might recall incorrectly).

You might have changed my mind, I no longer think that it's a bad idea :D

I've posted it to the upstream issue https://gitlab.gnome.org/GNOME/vte/-/issues/2495, let's see what others (most importantly VTE's main developer) thinks.

**Comment by antoine2711** [comment]
> Fair point, I vaguely recall (from ages ago) that VMware had such menu entries (of course I might recall incorrectly).

Yes, they also have that in VMs. I use VirtualBox, and it automatically syncs the clipboard with them. But, the security concerns in VMs are much lower, as usually, you are the sole user of the VMs hosted on your machine.

I'll read that discussion.

Thanks, and regards, Antoine
